shadowsocks/stream.go

```diff
@@ -18,15 +18,27 @@ import (
 	"bytes"
 	"crypto/cipher"
 	"crypto/rand"
+	"errors"
 	"fmt"
 	"io"
 
+	"github.com/Jigsaw-Code/outline-ss-server/shadowsocks/slicepool"
 	"github.com/shadowsocks/go-shadowsocks2/shadowaead"
 )
 
 // payloadSizeMask is the maximum size of payload in bytes.
 const payloadSizeMask = 0x3FFF // 16*1024 - 1
 
+// Maximum allowed cipher overhead
+const maxCipherOverhead = 16
+
+// The largest buffer we could need is for decrypting a max-length payload in
+// shadowsocksWriter.
+const maxBufferSize = 2 + maxCipherOverhead + payloadSizeMask + maxCipherOverhead
```

Comment:
Maybe rename this to `maxEncryptedChunkSize` to match the language in https://shadowsocks.org/en/spec/AEAD-Ciphers.html. It also conveys the semantic better.